### PR TITLE
Adds dynamodb sdk v1 backfill on index

### PR DESCRIPTION
**Pull request by jsibbison-square**
Allows the backfill to specify an indexName to be used in the scan rather than the main table.

### CODE REVIEW

File: client-dynamodb/build.gradle.kts [hunk 1]
```diff
@@ -52,6 +52,11 @@ dependencies {
       strictly("4.9.3")
     }
   }
+
+  // Required for M1 macs
+  if (org.apache.tools.ant.taskdefs.condition.Os.isArch("aarch64")) {
+    testImplementation("io.github.ganadist.sqlite4java:libsqlite4java-osx-aarch64:1.0.392")
```
Comment (mpawliszyn) [inline client-dynamodb/build.gradle.kts:58]: I don't think anyone else needs this? I wonder why this came up?

File: client-dynamodb/src/main/kotlin/app/cash/backfila/client/dynamodb/DynamoDbBackfill.kt [hunk 2]
```diff
@@ -76,4 +76,7 @@ abstract class DynamoDbBackfill<I : Any, P : Any> : Backfill {
 
   /** See [ScanRequest.setExpressionAttributeNames]. */
   open fun expressionAttributeNames(config: BackfillConfig<P>): Map<String, String>? = null
+
+  /** See [ScanRequest.setIndexName]. */
+  open fun indexName(config: BackfillConfig<P>): String? = null
```
Comment (mpawliszyn) [inline client-dynamodb/src/main/kotlin/app/cash/backfila/client/dynamodb/DynamoDbBackfill.kt:81]: I wonder if I should encapsulate all these arguments? Builder method might be nice and clear too.

File: client-dynamodb/src/test/kotlin/app/cash/backfila/client/dynamodb/DynamoDbIndexTest.kt [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+  private fun rowsInIndex(): List<TrackItem> {
+    val scanRequest = DynamoDBScanExpression().apply {
+      limit = 10000
+      indexName = "trackTitleIndex"
+    }
+    return dynamoDb.scan(TrackItem::class.java, scanRequest)
+  }
+
+  class MakeTracksExplicitBackfill @Inject constructor(
```
Comment (mpawliszyn) [inline client-dynamodb/src/test/kotlin/app/cash/backfila/client/dynamodb/DynamoDbIndexTest.kt:66]: A bit of a nit I guess but I probably would have made this mark tracks as singles or add the album to the track name in this case than a straight up copy that might become confusing later.

**Review by mpawliszyn — COMMENTED** [review]
@shellderp any other thoughts?

File: client-dynamodb/build.gradle.kts (hunk 1 above)
Comment (jsibbison-square) [reply client-dynamodb/build.gradle.kts]: Removed until I can find some definitive answer.

File: client-dynamodb/src/test/kotlin/app/cash/backfila/client/dynamodb/DynamoDbIndexTest.kt (hunk 3 above)
Comment (jsibbison-square) [reply client-dynamodb/src/test/kotlin/app/cash/backfila/client/dynamodb/DynamoDbIndexTest.kt:66]: Updated it to a distinct backfill now.

**Review by mpawliszyn — APPROVED** [review]

File: client-dynamodb/src/test/kotlin/app/cash/backfila/client/dynamodb/DynamoDbIndexTest.kt [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    return dynamoDb.scan(TrackItem::class.java, scanRequest)
+  }
+
+  class MakeTracksAsSinglesBackfill @Inject constructor(
+    dynamoDb: DynamoDBMapper,
+  ) : UpdateInPlaceDynamoDbBackfill<TrackItem, MakeTracksAsSinglesBackfill.SingleParameters>(dynamoDb) {
+
+    override fun runOne(item: TrackItem, config: BackfillConfig<SingleParameters>): Boolean {
+      val trackTitle = item.track_title ?: return false
+      if (trackTitle.endsWith(" (Single)")) return false // Idempotent retry?
```
[the remark this answers is not in the document]
Comment (jsibbison-square) [reply client-dynamodb/src/test/kotlin/app/cash/backfila/client/dynamodb/DynamoDbIndexTest.kt]: Updated

**Pull request merged**